### PR TITLE
Adding Pattern Matching Template

### LINKED ISSUE #893: SSH keychain widget does not recognize pattern matching declarations
### Dev Home version

0.137.141.0

### Windows build number

10.0.22621.0

### Other software

* OpenSSH_for_Windows_8.6p1, LibreSSL 3.4.3

### Steps to reproduce the bug

1. Add a new host to your `~/.ssh/config` file with [pattern matching](https://en.wikibooks.org/wiki/OpenSSH/Pattern_Matching_in_OpenSSH_Configuration) in the configuration

- Here was mine:
```
Host ??.logonoff
  HostName %h.co
  IdentityFile ~/.ssh/id_ed25519
```

2. Add the SSH keychain widget and try to connect to the host

image

3. Notice that it errors out when trying to connect as you cannot fill in the pattern

image

### Expected result

Prompt me to fill in the pattern, perhaps via a dialog box

### Actual result

Dev Home will try to SSH to the address with the question marks which is not a valid address

### Included System Information

_No response_

### Included Extensions Information

_No response_

### PR DESCRIPTION
**Pull request by joadoumie**
…a pattern to connect based on ssh config patterns## Summary of the pull request

## References and relevant issues
#893 

## Detailed description of the pull request / Additional comments
This PR allows for users to fill in hosts that have pattern matching(*,?) in the /.ssh/config file. 

image

image

image

## Validation steps performed
Locally tested on a variety of hosts in config file that contain pattern matching.

## PR checklist
- [x] Closes #893 
- [ ] Tests added/passed
- [ ] Documentation updated

### CODE REVIEW

**Comment by joadoumie** [comment]
Will do some additional testing around edge cases, remove some extraneous logging information, then move this PR out of a draft state.

File: CoreWidgetProvider/CoreWidgetProvider.csproj [hunk 1]
```diff
@@ -41,6 +41,9 @@
     <Content Include="Widgets\Templates\SSHWalletTemplate.json">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </Content>
+	  <Content Include="Widgets\Templates\SSHWalletPatternMatching.json">
```
Comment (EricJohnson327) [inline CoreWidgetProvider/CoreWidgetProvider.csproj:44]: Please fix formatting to match above lines.

File: CoreWidgetProvider/CoreWidgetProvider.csproj [hunk 2]
```diff
@@ -66,4 +69,6 @@
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </Content>
   </ItemGroup>
+
+  <ProjectExtensions><VisualStudio><UserProperties widgets_4templates_4sshwalletpatternmatching_1json__JsonSchema="https://codemagic.io/codemagic-schema.json" /></VisualStudio></ProjectExtensions>
```
Comment (EricJohnson327) [inline CoreWidgetProvider/CoreWidgetProvider.csproj:73]: What is this?

Comment (joadoumie) [reply CoreWidgetProvider/CoreWidgetProvider.csproj:73]: I honestly don't know - I didn't see that in the git diff. I didn't manually put that there.

Comment (joadoumie) [reply CoreWidgetProvider/CoreWidgetProvider.csproj]: Removed it and project builds as expected.

File: CoreWidgetProvider/CoreWidgetProvider.csproj (hunk 1 above)
Comment (joadoumie) [reply CoreWidgetProvider/CoreWidgetProvider.csproj]: Fixed.

Comment (joadoumie) [reply CoreWidgetProvider/CoreWidgetProvider.csproj]: ok phew - I've fixed it.

File: CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  "type": "AdaptiveCard",
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "version": "1.5",
+  "body": [
+    {
+      "type": "Container",
+      "items": [
+        {
+          "type": "TextBlock",
+          "text": "Your SSH key contains pattern matching. Fill out the pattern to connect.",
```
Comment (stefansjfw) [inline CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json:11]: This should be localized

Comment (joadoumie) [reply CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json:11]: Good catch. Will add to resw and test.

Comment (joadoumie) [reply CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json:11]: How can I properly test if my reference to resw is working? I have added the string here: 

image

But when I debug the app it doesn't seem to be finding the correct string. Any tips for how to properly ensure I reference resw?

Comment (joadoumie) [reply CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json]: I did not notice the Resources.cs file - I have fixed it.

File: CoreWidgetProvider/Widgets/SSHWalletWidget.cs [hunk 4]
```diff
@@ -324,13 +353,33 @@ public override void UpdateWidget()
         WidgetManager.GetDefault().UpdateWidget(updateOptions);
     }
 
+    public void UpdateWidget(string patternHost)
```
Comment (krschau) [inline CoreWidgetProvider/Widgets/SSHWalletWidget.cs:356]: I think it's a little confusing to have this UpdateWidget which is kind of like the overridden UpdateWidget but also not. Can use the regular UpdateWidget(), and get the patternHost into the data in the GetData(WidgetPageState page) method?

File: CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    {
+      "type": "Container",
+      "items": [
+        {
+          "type": "TextBlock",
+          "text": "%SSH_Widget_Template/PatternMatching%",
+          "wrap": true,
+          "horizontalAlignment": "Left",
+          "size": "Medium",
+          "weight": "Bolder"
```
Comment (krschau) [inline CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json:15]: I'd love for Design to give this a quick peek before checking in if they haven't yet.

File: CoreWidgetProvider/Widgets/SSHWalletWidget.cs (hunk 4 above)
Comment (joadoumie) [reply CoreWidgetProvider/Widgets/SSHWalletWidget.cs:356]: Fair point! I'll look into it!

File: CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json (hunk 5 above)
Comment (joadoumie) [reply CoreWidgetProvider/Widgets/Templates/SSHWalletPatternMatching.json:15]: Will do.

File: CoreWidgetProvider/Widgets/SSHWalletWidget.cs [hunk 6]
```diff
@@ -111,16 +114,38 @@ public override void OnActionInvoked(WidgetActionInvokedArgs actionInvokedArgs)
...
+            {
+                data = patternHostNode.ToString();
+            }
+        }
+
+        if (data.Contains('*') || data.Contains('?'))
+        {
+            Page = WidgetPageState.Pattern;
+            UpdateWidget(data);
+            return;
```
[the remark this answers is not in the document]
Comment (joadoumie) [reply CoreWidgetProvider/Widgets/SSHWalletWidget.cs:146]: Will add comments thanks for the feedback.

File: CoreWidgetProvider/Widgets/SSHWalletWidget.cs [hunk 7]
```diff
@@ -36,6 +37,8 @@ public SSHWalletWidget()
 
     ~SSHWalletWidget()
     {
+        // Ensures widget is in the proper view when a user returns
+        UpdateWidget();
```
[the remark this answers is not in the document]
Comment (joadoumie) [reply CoreWidgetProvider/Widgets/SSHWalletWidget.cs:41]: Is the destructor only called on the 'remove widget' event? I was trying to ensure that if a user gets into a pattern matching state but didn't end up connecting to an SSH connection that if they were to close Dev Home and return the json that is being shown is the original one with the list of ssh connections rather than the pattern matching template. If the destructor is only called on the 'Remove Widget' action then I think this is irrelevant and I will just remove it.

[the remark this answers is not in the document]
Comment (krschau) [reply CoreWidgetProvider/Widgets/SSHWalletWidget.cs:41]: I expect that the next time the constructor is called for this widget, it will be created in the right state, so this shouldn't be needed here. If that's not the case, probably a thing to fix :)

**Comment by joadoumie** [comment]
closing out this PR and will create a new updated one with the requested changes. A lot has changed since this PR was originally created so it seems easiest to create a new PR to resolve merge conflicts
